python/mxnet/profiler.py

```diff
@@ -56,3 +69,237 @@ def dump_profile():
+def tune_pause():
+    check_call(_LIB.MXProfileTunePause())
+
+def tune_resume():
+    check_call(_LIB.MXProfileTuneResume())
+
+def create_counter(domain_handle, name, value=None):
+    counter_handle = ProfileCounterHandle()
+    check_call(_LIB.MXProfileCreateCounter(domain_handle,
+                                           c_str(name),
+                                           ctypes.byref(counter_handle)))
+    if value is not None:
+        set_counter(counter_handle, value)
+    return counter_handle
+
+def destroy_counter(counter_handle):
+    check_call(_LIB.MXProfileDestroyCounter(counter_handle))
+
+def set_counter(counter_handle, value):
+    check_call(_LIB.MXProfileSetCounter(counter_handle, int(value)))
+
+def increment_counter(counter_handle, by_value):
+    check_call(_LIB.MXProfileAdjustCounter(counter_handle, int(by_value)))
+
+def decrement_counter(counter_handle, by_value):
+    check_call(_LIB.MXProfileAdjustCounter(counter_handle, -int(by_value)))
+
+def set_append_mode(mode):
+  if mode is False:
+    mode = 0
+  else:
+    mode = 1
+  check_call(_LIB.MXSetDumpProfileAppendMode(int(mode)))
+
+def set_continuous_dump(continuous_dump=True, delay_in_seconds=1.0):
+  if continuous_dump is False:
+    cd = 0
+  else:
+    cd = 1
+  ds = float(delay_in_seconds)
+  check_call(_LIB.MXSetContinuousProfileDump(ctypes.c_int(cd), ctypes.c_float(ds)))
+
+def set_instant_marker(domain_handle, name, scope='process'):
+    marker_scope2int = { 'global': 1, 'process': 2, 'thread': 3, 'task': 4, 'marker': 5 }
+    scope_int = marker_scope2int[scope]
+    check_call(_LIB.MXProfileSetInstantMarker(domain_handle, c_str(name), scope_int))
+
+
+class Domain:
+    """Profiling domain, used to group sub-objects like tasks, counters, etc into categories
+    Serves as part of 'categories' for chrome://tracing
+    Note: Domain handles are never destroyed
+    """
+    def __init__(self, name):
+        self.name = name
+        self.handle = create_domain(name)
+
+    def __str__(self):
+        return self.name
+
+
+class Task:
+    """Profiling Task class
+    A task is a logical unit of work performed by a particular thread.
+    Tasks can nest; thus, tasks typically correspond to functions, scopes, or a case block
+    in a switch statement.
+    You can use the Task API to assign tasks to threads
+    """
+    def __init__(self, domain, name):
+        self.domain = domain
+        self.name = name
+        self.handle = create_task(domain.handle, name)
+
+    def start(self):
+        task_start(self.handle)
+
+    def stop(self):
+        task_stop(self.handle)
+
+    def __str__(self):
+        return self.name
+
+    def __del__(self):
+        if self.handle is not None:
+            destroy_task(self.handle)
+
+
+class Frame:
+    """Profiling Frame class
+    Use the frame API to insert calls to the desired places in your code and analyze
+    performance per frame, where frame is the time period between frame begin and end points.
```

Comment:
what's the difference between frame and task?

Reply:
Frame is displayed differently in VTune. It is displayed at the bottom with a "frame rate". I figure it's meant for recurring things, such as "batch rate" in our case.
Tasks and Frames can also be separated by different "Grouping" by the selection of the drop box up top.  That drop box has all sorts of combinations of domains, tasks, frames, etc. which changes what you see in the main window.

Reply:
Still don't quite understand. 
This sounds very vtune specific.

Reply:
It’s a mapped to a vtune object, yes. In chrome tracing it will just show as a reentrant duration same as a “task”. If you don’t have vtune then you probably won’t use it.

Reply:
We should do something more generic than just mapping to vtune concepts.
It's very hard to understand what these are about for people unfamiliar with vtune

Reply:
Btw why do we expose this to the front end?

Reply:
Why do we expose what? Profiling objects?

Reply:
I think the vtune naming is pretty generic. No one is forcing people to use it. Frame can actually be implemented more closely in chrome tracing if paired with a counter to show “frame rate”. I can do that in a later PR

Reply:
I can name them something else, if you insist but consider:
Vtune has set labels and grouping and drop-downs and well-documented behavior for these objects. Chrome tracing doesn’t really have any visual object naming or grouping, so naming them something non-vtune-ish doesn’t really buy you anything extra with chrome and makes the use of
Vtune totally confusing because you have to use guesswork to understand what’s what in the UI.